src/Autofac/Core/Registration/ComponentRegistryBuilder.cs

```diff
@@ -0,0 +1,184 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+
+using Autofac.Builder;
+using Autofac.Util;
+
+namespace Autofac.Core.Registration
+{
+    /// <summary>
+    /// Used to build a <see cref="IComponentRegistry" />.
+    /// </summary>
+    internal class ComponentRegistryBuilder : Disposable, IComponentRegistryBuilder
+    {
+        /// <summary>
+        /// Protects instance variables from concurrent access.
+        /// </summary>
+        private readonly object _synchRoot = new object();
+
+        /// <summary>
+        /// The tracker for the registered services.
+        /// </summary>
+        private readonly IRegisteredServicesTracker _registeredServicesTracker;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ComponentRegistryBuilder"/> class.
+        /// </summary>
+        /// <param name="registeredServicesTracker">The tracker for the registered services.</param>
+        /// <param name="properties">The properties used during component registration.</param>
+        internal ComponentRegistryBuilder(IRegisteredServicesTracker registeredServicesTracker, IDictionary<string, object> properties)
+        {
+            Properties = properties;
+            _registeredServicesTracker = registeredServicesTracker;
+        }
+
+        /// <summary>
+        /// Releases unmanaged and - optionally - managed resources.
+        /// </summary>
+        /// <param name="disposing"><c>true</c> to release both managed and unmanaged resources; <c>false</c> to release only unmanaged resources.</param>
+        protected override void Dispose(bool disposing)
+        {
+            _registeredServicesTracker.Dispose();
+
+            base.Dispose(disposing);
+        }
+
+        /// <summary>
+        /// Gets the set of properties used during component registration.
+        /// </summary>
+        /// <value>
+        /// An <see cref="IDictionary{TKey, TValue}"/> that can be used to share
+        /// context across registrations.
+        /// </value>
+        public IDictionary<string, object> Properties { get; }
```

Comment:
For example, regarding synchronization, `Properties` here will allow the underlying `Dictionary<,>` to be accessed outside the `_syncRoot` lock, which is potentially non-threadsafe.

(I'm happy to attempt a PR for these things but just thought I should confirm my understanding first.)

Reply:
That is true. I kept this as it is now. This is something that could be improved.